dnn/utils/yolo_encoder.py

```diff
@@ -0,0 +1,138 @@
+        
+        # That means no box of this size
+        if (n == 0):
+            return y_true_out_gg36
+
+        #box shape will be (n, 4)
+        box_n4 = y_true_n6[...,0:4]
+        box_yx = (y_true_n6[...,0:2] + y_true_n6[...,2:4]) / 2
+        grid_yx = tf.cast(box_yx // (1/grid_size), tf.int32)
+
+        #index should be contain info regarding which box to update
+        #shape (n,3)
+        #e.g.: [7,7,0] means grid 7*7 and anchor 0
+        indexes_n3 = tf.stack(
+            [grid_yx[...,0], 
+            grid_yx[...,1], 
+            tf.cast(y_true_n6[...,5] % 3, 
+            dtype=tf.int32)], axis=1)
+        
+        #update should contain object info, size (n, 6)
+        #(x, y, w, h, obj, class) at pos shape
+        updates_n6 = tf.stack(
+            [box_n4[...,1], 
+            box_n4[...,0], 
+            box_n4[...,2]-box_n4[...,0], 
+            box_n4[...,3]-box_n4[...,1], 
+            tf.ones((n,)), 
+            y_true_n6[...,4]], axis=1)
+
+        out = tf.tensor_scatter_nd_update(y_true_out_gg36, indexes_n3, updates_n6)
+        return out
+
+
+    def transform_label(self, y_train_n5):
+        '''
+        Read raw y_train and return yolo_out of different dimensions.
+
+        args:
+            y_train_n5: label, tensor of shape (n, (x,y,x,y,class)), should be in range(0, 1)
+        '''
+        y_outs = []
+        grid_size = self.size // 32
+
+        # calculate anchor index for true boxes
+        anchors_n2 = tf.cast(self.anchors, tf.float32)
+        anchor_area = anchors_n2[..., 0] * anchors_n2[..., 1]
+
+        # y_true is yxyx after imgaug
+        # How to calculate hw depends on tf_record format
+        box_hw = y_train_n5[..., 2:4] - y_train_n5[..., 0:2]
+        box_hw = tf.tile(tf.expand_dims(box_hw, -2),
+                        (1, tf.shape(anchors_n2)[0], 1))
+        box_area = box_hw[..., 0] * box_hw[..., 1]
+
+        # Calculate intersection between bbox and anchors
+        intersection = tf.minimum(box_hw[..., 1], anchors_n2[..., 0]) * \
+            tf.minimum(box_hw[..., 0], anchors_n2[..., 1])
+
+        # Calculate iou, which is intersection/total area
+        iou = intersection / (box_area + anchor_area - intersection)
+        
+        # Find the best anchor
+        anchor_idx = tf.cast(tf.argmax(iou, axis=-1), tf.float32)
+        anchor_idx = tf.expand_dims(anchor_idx, axis=-1)
+
+        # n*6
+        y_train_n6 = tf.concat([y_train_n5, anchor_idx], axis=-1)
+
+        for masks in self.anchor_masks:
+            y_outs.append(self.transform_label_for_output(
+                y_train_n6, grid_size, masks))
+            grid_size *= 2
+        
+        #y_out[0] = 13*13*3*6, y_out[1] = 26*26*3*6, y_out[2] = 52*52*3*6
+        # grid*grid*anchor*(x,y,w,h,confidence,class(2))
+        return tuple(y_outs)
+
+    #Placeholder for lambda function
+    def transform_images_train(self, x_train):
+        return x_train
```

Comment:
What is this for?

Reply:
This is kind of a placeholder function for lambda x,y I guess, e.g.
``` python
train = loadDataset("whatever.tfrecords")
train.map(lambda: x, y : (
        dataset.transform_images(x), # Not sure if I can do just x
        dataset.transform_label(y)
))
```

Reply:
I see. That makes sense. Maybe it's more straightforward to make a function like `encode_dataset(self, raw_dataset)` and return a "mapped" dataset object. Otherwise, it might not seem obvious on the user side to use those helpers in such ways.